utils/_context/_scenarios/core.py

```diff
@@ -310,13 +311,62 @@ def _get_warmups(self):
         if not self.replay:
             warmups.append(create_network)
 
-            for container in self._required_containers:
-                warmups.append(container.start)
+        # on replay mode, _start_containers() won't actually start containers,
+        # but will only call the necessary post_start method
+        warmups.append(self._start_containers)
+
+        return warmups
+
+    def _start_containers(self):
+        dependencies = self._get_dependencies()
```

Comment:
The `dependencies` name is definitly not self-explanatory. We cannot understand what is this data structure without reading the code that produces it. 

`_get_dependencies` misleading me, as by its name, I initially understood it was a list of deps, where it's more a map.

`_start_containers` is not as simple as "start container". It recursively crawls the dependency tree. In particular, if I'm not wrong, the algo is  sensible to the order we add container.

Reply:
I added clarifying comments, let me know if that works!

> In particular, if I'm not wrong, the algo is sensible to the order we add container.

This is no longer the case, it can be in any order.

Reply:
Let say you have 

```yml
parent_1:
   depends_on: [child_A]
parent_2:
   depends_on: [child_B]
parent_3:
   depends_on: [child_C]
parent_4:
   depends_on: [child_A, child_B, child_C]
```

It'll start sequentially `child_A`, `parent_1`, `child_B`, `parent_2`, `child_C`, `parent_3`, then `parent_4`.

But if you have : 


```yml
parent_4:
   depends_on: [child_A, child_B, child_C]
parent_1:
   depends_on: [child_A]
parent_2:
   depends_on: [child_B]
parent_3:
   depends_on: [child_C]
```

if I'm not wrong, it'll start in parallel `child_A`, `child_B`, `child_C`, then in parallel `parent_4`, `parent_1`, `parent_2` and `parent_3`, right ? 

So far, I'm fine with that, let's not do over-engineering. But if it's true, I think having a snall comment somewhere explaining that for performance reason, containers with lot of common depens should be placed in first will help a lot.

Reply:
> if I'm not wrong, it'll start in parallel child_A, child_B, child_C, then in parallel parent_4, parent_1, parent_2 and parent_3, right ?

No, both cases will work as your first example. This is why there are 2 loops, one to build the dependency tree, and another one to start the containers according to the dependencies that have been calculated by looking at every container definition first.

Reply:
> No, both cases will work as your first example. 

As my **first** example, so the worst case whatever the order in the list ? 🤔 


> This is why there are 2 loops, one to build the dependency tree,

Then could you add more comments about the data structure produced by `_get_dependencies`?